src/transaction/log_replication.cpp

```diff
@@ -286,6 +288,7 @@ namespace cublog
 	  // however, this would need one more mutex lock; therefore, suffice to do it here
 	  assert (m_replication_active);
 
+	  m_processed_lsa = m_redo_lsa;
```

Comment:
Thank you for the detailed feedback :).
 With m_proccessed_lsa only in the atomic_replicator, I'm concerned that `get_highest_processed_lsa()` will return different value which has different meanings according to its class. The replicator will return m_redo_lsa, which will be processed, and the atomic replicator will return m_processed_lsa, which has been processed. I think the same interface should have the same semantics.
 On top of that, get_highest_processed_lsa() already returns the lsa that is not processed actually (m_redo_lsa). So, I don't understand why we can't use m_processed_lsa instead of m_redo_lsa. If the name is the problem, we can rename the variable. Do I miss or misunderstand something?

Reply:
You are correct that base and overridden functions should implement the same semantics.
And you are correct that the current implementation is - well - inconsistent to say the least.
What about this:
- `get_highest_processed_lsa` is only called on PTS; never on PS:
  - in `pgbuf_wait_for_replication` on PTS
  - in `pgbuf_read_page_from_file_or_page_server` only for PTS as well
- and still override `get_highest_processed_lsa` in atomic replicator and implement with the newly added member `m_processed_lsa` which will be correct
- this seems to be most constrained to the scenarios that are currently needed

```
  log_lsa
  replicator::get_highest_processed_lsa () const
  {
    assert_release (false);
  }

  log_lsa
  replicator::get_lowest_unapplied_lsa () const
  {
    // TODO: needs to be refactored to work with the new replicators flavors
    if (m_parallel_replication_redo == nullptr)
      {
	// sync
        std::lock_guard<std::mutex> lockg (m_redo_lsa_mutex);
        return m_redo_lsa;
      }

    // a different value will return from here when the atomic replicator is added
    // for now this part should not be reached
    assert_release (false);
  }
```